appstream stack resource has issue with changing user_settings and re-running terraform apply

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform CLI and Terraform AWS Provider Version

```hcl
terraform {
  required_providers {
    aws = {
      source = "hashicorp/aws"
      version = "3.73.0"
    }
  }
}
```

### Affected Resource(s)

- aws_appstream_stack

### Terraform Configuration Files

```hcl
resource "aws_appstream_fleet" "test_fleet" {
  name          = "cc-tmp-fleet-1"
  display_name  = "cc-test-appstream-fleet"
  image_name    = "Amazon-AppStream2-Sample-Image-02-04-2019"
  instance_type = "stream.standard.small"
  stream_view   = "DESKTOP"
  enable_default_internet_access = true
  fleet_type    = "ON_DEMAND"

  vpc_config {
    security_group_ids = [aws_security_group.appstream_sec_group.id]
    subnet_ids = data.aws_subnet_ids.cc_subnets.ids
  } 

  compute_capacity {
    desired_instances = 1
  }

  depends_on = [aws_iam_role.appstream_role]
}

resource "aws_appstream_stack" "test_stack" {
  name = "cc-tmp-stack-1"
  display_name = "cc-test-appstream-stack"

  user_settings {
    action     = "CLIPBOARD_COPY_FROM_LOCAL_DEVICE"
    permission = "DISABLED"
  }
  user_settings {
    action     = "CLIPBOARD_COPY_TO_LOCAL_DEVICE"
    permission = "DISABLED"
  }
  user_settings {
    action     = "FILE_UPLOAD"
    permission = "ENABLED"
  }
  user_settings {
    action     = "FILE_DOWNLOAD"
    permission = "DISABLED"
  }

  storage_connectors {
    connector_type = "HOMEFOLDERS"
  }

  depends_on = [aws_iam_role.appstream_role]
}
```

### Expected Behavior

The settings should have been changed successfully.

### Actual Behavior

Received provider errors below:

```plaintext
 Error: Provider produced inconsistent final plan
│
│ When expanding the plan for aws_appstream_stack.test_stack to include new values learned so far during apply, provider "registry.terraform.io/hashicorp/aws" produced an invalid new value for .user_settings: planned set element cty.ObjectVal(map[string]cty.Value{"action":cty.StringVal("CLIPBOARD_COPY_TO_LOCAL_DEVICE"),       
│ "permission":cty.StringVal("DISABLED")}) does not correlate with any element in actual.
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent final plan
│
│ When expanding the plan for aws_appstream_stack.test_stack to include new values learned so far during apply, provider "registry.terraform.io/hashicorp/aws" produced an invalid new value for .user_settings: planned set element cty.ObjectVal(map[string]cty.Value{"action":cty.StringVal("FILE_DOWNLOAD"),
│ "permission":cty.StringVal("DISABLED")}) does not correlate with any element in actual.
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent final plan
│
│ When expanding the plan for aws_appstream_stack.test_stack to include new values learned so far during apply, provider "registry.terraform.io/hashicorp/aws" produced an invalid new value for .user_settings: planned set element cty.ObjectVal(map[string]cty.Value{"action":cty.StringVal("FILE_UPLOAD"),
│ "permission":cty.StringVal("")}) does not correlate with any element in actual.
│
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
```

### Steps to Reproduce

Run an initial terraform apply and then change the settings and try re-running a terraform apply

Any future applies I try now are crashing the plugins. Seems there are a few different bugs that people are encountering when using appstream resources :(

╷
│ Error: Provider produced inconsistent final plan
│ 
│ When expanding the plan for module.fintech_standard_infrastructure.module.appstream-access[0].aws_appstream_stack.example_stack to include new values
│ learned so far during apply, provider "registry.terraform.io/hashicorp/aws" produced an invalid new value for .user_settings: actual set element
│ cty.ObjectVal(map[string]cty.Value{"action":cty.StringVal("FILE_DOWNLOAD"), "permission":cty.StringVal("DISABLED")}) does not correlate with any
│ element in plan.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent final plan
│ 
│ When expanding the plan for module.fintech_standard_infrastructure.module.appstream-access[0].aws_appstream_stack.example_stack to include new values
│ learned so far during apply, provider "registry.terraform.io/hashicorp/aws" produced an invalid new value for .user_settings: actual set element
│ cty.ObjectVal(map[string]cty.Value{"action":cty.StringVal("PRINTING_TO_LOCAL_DEVICE"), "permission":cty.StringVal("DISABLED")}) does not correlate with
│ any element in plan.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Provider produced inconsistent final plan
│ 
│ When expanding the plan for module.fintech_standard_infrastructure.module.appstream-access[0].aws_appstream_stack.example_stack to include new values
│ learned so far during apply, provider "registry.terraform.io/hashicorp/aws" produced an invalid new value for .user_settings: block set length changed
│ from 3 to 5.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
╷
│ Error: Plugin did not respond
│ 
│   with module.fintech_standard_infrastructure.module.appstream-access[0].aws_appstream_fleet.example_fleet,
│   on .terraform/modules/standard_infrastructure.appstream-access/appstream.tf line 3, in resource "aws_appstream_fleet" "example_fleet":
│    3: resource "aws_appstream_fleet" "example_fleet" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more
│ details.
╵
╷
│ Error: Plugin did not respond
│ 
│   with module.standard_infrastructure.module.appstream-access[0].aws_appstream_fleet.example2_fleet,
│   on .terraform/modules/fintech_standard_infrastructure.appstream-access/appstream.tf line 71, in resource "aws_appstream_fleet" "example2_fleet":
│   71: resource "aws_appstream_fleet" "example2_fleet" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more
│ details.
╵
╷
│ Error: Plugin did not respond
│ 
│   with module.standard_infrastructure.module.appstream-access[0].aws_appstream_stack.example2_stack,
│   on .terraform/modules/standard_infrastructure.appstream-access/appstream.tf line 93, in resource "aws_appstream_stack" "example2_stack":
│   93: resource "aws_appstream_stack" "example2_stack" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more
│ details.
╵
╷
│ Error: Plugin did not respond
│ 
│   with module.fintech_standard_infrastructure.aws_s3_bucket.local_bucket,
│   on .terraform/modules/standard_infrastructure/s3_buckets.tf line 4, in resource "aws_s3_bucket" "local_bucket":
│    4: resource "aws_s3_bucket" "local_bucket" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more
│ details.
╵
╷
│ Error: Plugin did not respond
│ 
│   with module.fintech_standard_infrastructure.aws_s3_bucket.waf_logs_bucket,
│   on .terraform/modules/standard_infrastructure/s3_buckets.tf line 66, in resource "aws_s3_bucket" "waf_logs_bucket":
│   66: resource "aws_s3_bucket" "waf_logs_bucket" {
│ 
│ The plugin encountered an error, and failed to respond to the plugin.(*GRPCProvider).UpgradeResourceState call. The plugin logs may contain more
│ details.
╵
Releasing state lock. This may take a few moments...

Stack trace from the terraform-provider-aws_v3.73.0_x5 plugin:

panic: interface conversion: interface {} is []interface {}, not *schema.Set

goroutine 121 [running]:
github.com/hashicorp/terraform-provider-aws/internal/service/appstream.resourceStackUpdate(0xa4cab58, 0xc001fdf7c0, 0xc001fe9a00, 0x8234060, 0xc000f98000, 0xc00200b6d0, 0xabec2a, 0xc001fe9880)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-aws/internal/service/appstream/stack.go:338 +0xfad
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).update(0xc0007b80e0, 0xa4cab58, 0xc001fdf7c0, 0xc001fe9a00, 0x8234060, 0xc000f98000, 0x0, 0x0, 0x0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.1/helper/schema/resource.go:370 +0xa7
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).Apply(0xc0007b80e0, 0xa4cab58, 0xc001fdf7c0, 0xc001df76c0, 0xc001fe9880, 0x8234060, 0xc000f98000, 0x0, 0x0, 0x0, ...)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.1/helper/schema/resource.go:475 +0x390
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ApplyResourceChange(0xc000288ff0, 0xa4cab58, 0xc001fdf7c0, 0xc001df5e50, 0x956ff89, 0x12, 0x0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.1/helper/schema/grpc_provider.go:977 +0xacf
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ApplyResourceChange(0xc000290b00, 0xa4cac00, 0xc001fdf7c0, 0xc001c8dc00, 0x0, 0x0, 0x0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.5.0/tfprotov5/tf5server/server.go:603 +0x465
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x92d84a0, 0xc000290b00, 0xa4cac00, 0xc00204e300, 0xc001fe57a0, 0x0, 0xa4cac00, 0xc00204e300, 0xc0016c2880, 0xc84)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.5.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:380 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0002b3a40, 0xa4f75b8, 0xc001c72180, 0xc000294600, 0xc0018fcb40, 0xf05eb80, 0x0, 0x0, 0x0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1194 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc0002b3a40, 0xa4f75b8, 0xc001c72180, 0xc000294600, 0x0)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1517 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000e24de0, 0xc0002b3a40, 0xa4f75b8, 0xc001c72180, 0xc000294600)
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:859 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd

Error: The terraform-provider-aws_v3.73.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

--- a/.changelog/28766.txt
+++ b/.changelog/28766.txt
@@ -1,0 +1,7 @@
+```release-note:bug
+resource/aws_appstream_stack: Fix panic on user_settings update
+```
+
+```release-note:bug
+resource/aws_appstream_stack: Prevent unnecessary replacements on update
+```

--- a/internal/service/appstream/stack.go
+++ b/internal/service/appstream/stack.go
@@ -25,10 +25,6 @@ import (
 	"github.com/hashicorp/terraform-provider-aws/internal/verify"
 )
 
-var (
-	flagDiffUserSettings = false
-)
-
 func ResourceStack() *schema.Resource {
 	return &schema.Resource{
 		CreateWithoutTimeout: resourceStackCreate,
@@ -707,6 +703,7 @@ func flattenUserSettings(apiObjects []*appstream.UserSetting) []map[string]inter
 }
 
 func suppressAppsStreamStackUserSettings(k, old, new string, d *schema.ResourceData) bool {
+	flagDiffUserSettings := false
 	count := len(d.Get("user_settings").(*schema.Set).List())
 	defaultCount := len(appstream.Action_Values())
 

--- a/internal/service/appstream/stack.go
+++ b/internal/service/appstream/stack.go
@@ -402,7 +402,7 @@ func resourceStackUpdate(ctx context.Context, d *schema.ResourceData, meta inter
 	}
 
 	if d.HasChange("user_settings") {
-		input.UserSettings = expandUserSettings(d.Get("user_settings").([]interface{}))
+		input.UserSettings = expandUserSettings(d.Get("user_settings").(*schema.Set).List())
 	}
 
 	resp, err := conn.UpdateStack(input)

--- a/internal/service/appstream/stack.go
+++ b/internal/service/appstream/stack.go
@@ -91,15 +91,11 @@ func ResourceStack() *schema.Resource {
 			"description": {
 				Type:         schema.TypeString,
 				Optional:     true,
-				Computed:     true,
-				ForceNew:     true,
 				ValidateFunc: validation.StringLenBetween(0, 256),
 			},
 			"display_name": {
 				Type:         schema.TypeString,
 				Optional:     true,
-				Computed:     true,
-				ForceNew:     true,
 				ValidateFunc: validation.StringLenBetween(0, 100),
 			},
 			"embed_host_domains": {
@@ -184,7 +180,7 @@ func ResourceStack() *schema.Resource {
 				},
 				Set: userSettingsHash,
 			},
-			"tags":     tftags.TagsSchemaForceNew(),
+			"tags":     tftags.TagsSchema(),
 			"tags_all": tftags.TagsSchemaComputed(),
 		},
 

--- a/internal/service/appstream/stack_test.go
+++ b/internal/service/appstream/stack_test.go
@@ -110,7 +110,7 @@ func TestAccAppStreamStack_complete(t *testing.T) {
 					resource.TestCheckResourceAttr(resourceName, "feedback_url", ""),
 					resource.TestCheckResourceAttr(resourceName, "redirect_url", ""),
 					resource.TestCheckResourceAttr(resourceName, "storage_connectors.#", "1"),
-					resource.TestCheckResourceAttr(resourceName, "user_settings.#", "4"),
+					resource.TestCheckResourceAttr(resourceName, "user_settings.#", "7"),
 					resource.TestCheckResourceAttr(resourceName, "tags.%", "0"),
 					resource.TestCheckResourceAttr(resourceName, "tags_all.%", "0"),
 				),
@@ -380,19 +380,28 @@ resource "aws_appstream_stack" "test" {
     action     = "CLIPBOARD_COPY_FROM_LOCAL_DEVICE"
     permission = "ENABLED"
   }
-
   user_settings {
     action     = "CLIPBOARD_COPY_TO_LOCAL_DEVICE"
     permission = "ENABLED"
   }
-
+  user_settings {
+    action     = "DOMAIN_PASSWORD_SIGNIN"
+    permission = "ENABLED"
+  }
+  user_settings {
+    action     = "DOMAIN_SMART_CARD_SIGNIN"
+    permission = "DISABLED"
+  }
+  user_settings {
+    action     = "FILE_DOWNLOAD"
+    permission = "ENABLED"
+  }
   user_settings {
     action     = "FILE_UPLOAD"
     permission = "ENABLED"
   }
-
   user_settings {
-    action     = "FILE_DOWNLOAD"
+    action     = "PRINTING_TO_LOCAL_DEVICE"
     permission = "ENABLED"
   }
 
@@ -451,29 +460,28 @@ resource "aws_appstream_stack" "test" {
     action     = "CLIPBOARD_COPY_FROM_LOCAL_DEVICE"
     permission = "ENABLED"
   }
-
   user_settings {
     action     = "CLIPBOARD_COPY_TO_LOCAL_DEVICE"
     permission = "ENABLED"
   }
-
   user_settings {
-    action     = "FILE_UPLOAD"
+    action     = "DOMAIN_PASSWORD_SIGNIN"
+    permission = "ENABLED"
+  }
+  user_settings {
+    action     = "DOMAIN_SMART_CARD_SIGNIN"
     permission = "DISABLED"
   }
-
   user_settings {
     action     = "FILE_DOWNLOAD"
     permission = "ENABLED"
   }
-
   user_settings {
-    action     = "PRINTING_TO_LOCAL_DEVICE"
+    action     = "FILE_UPLOAD"
     permission = "ENABLED"
   }
-
   user_settings {
-    action     = "DOMAIN_PASSWORD_SIGNIN"
+    action     = "PRINTING_TO_LOCAL_DEVICE"
     permission = "ENABLED"
   }
 

--- a/website/docs/r/appstream_stack.html.markdown
+++ b/website/docs/r/appstream_stack.html.markdown
@@ -33,11 +33,23 @@ resource "aws_appstream_stack" "example" {
     permission = "ENABLED"
   }
   user_settings {
+    action     = "DOMAIN_PASSWORD_SIGNIN"
+    permission = "ENABLED"
+  }
+  user_settings {
+    action     = "DOMAIN_SMART_CARD_SIGNIN"
+    permission = "DISABLED"
+  }
+  user_settings {
+    action     = "FILE_DOWNLOAD"
+    permission = "ENABLED"
+  }
+  user_settings {
     action     = "FILE_UPLOAD"
     permission = "ENABLED"
   }
   user_settings {
-    action     = "FILE_DOWNLOAD"
+    action     = "PRINTING_TO_LOCAL_DEVICE"
     permission = "ENABLED"
   }
 
@@ -71,7 +83,7 @@ The following arguments are optional:
 * `redirect_url` - (Optional) URL that users are redirected to after their streaming session ends.
 * `storage_connectors` - (Optional) Configuration block for the storage connectors to enable.
   See [`storage_connectors`](#storage_connectors) below.
-* `user_settings` - (Optional) Configuration block for the actions that are enabled or disabled for users during their streaming sessions. By default, these actions are enabled.
+* `user_settings` - (Optional) Configuration block for the actions that are enabled or disabled for users during their streaming sessions. If not provided, these settings are configured automatically by AWS. If provided, the terraform configuration should include a block for each configurable action.
   See [`user_settings`](#user_settings) below.
 * `tags` - (Optional) Key-value mapping of resource tags. If configured with a provider [`default_tags` configuration block](https://registry.terraform.io/providers/hashicorp/aws/latest/docs#default_tags-configuration-block) present, tags with matching keys will overwrite those defined at the provider-level.
 

